src/include/gcc.h

```diff
@@ -89,38 +89,29 @@
  * swap) operations.
  */
 
-#ifdef __clang__
-/*
- * We avoid __sync_bool_compare_and_swap with due to problems with optimization
```

Comment:
@keithbostic 
Spooky stuff right here...

Ideally, I'd like to remove this special case since we're not using that set of intrinsics anymore. But to my understanding, the `__sync` series of builtins is implemented in terms of the more flexible `__atomic` ones. So it may suffer from the same problem.

If you can remember the story behind this, could you please fill me in on:
- Whether this was "obvious" and caused build failures often.
- What part of WiredTiger exposed this bug. I see that in the issue tracker, the LLVM dev who responded couldn't reproduce so I'm assuming this "optimisation" doesn't always trigger. If I'm going to look at what's being emitted by Clang then I'll need to narrow down to a usage that is known to cause this.

If the bug was obvious in testing, I'd be tempted to just merge this and see if it breaks anything in Jenkins.

Reply:
While it's been 4.5 years and 5 major clang releases since we saw the problem, I'm hesitant to remove the workaround. Further, we don't know when (well, honestly, "if"), the bug has been fixed, so we can't even detect versioning problems at compile time. (As an example, MongoDB/Evergreen allows you to spawn Ubuntu 1404, which has clang 3.4 installed.) 

I don't know how difficult it was to find or reproduce, @michaelcahill did that work.

cc: @agorrod

Reply:
The workaround is a blocker since it is Clang who is noisily complaining about our use of `__sync` functions. So I can't have:
```
#ifdef __clang__
/* __sync versions of the functions */
#else
/* __atomic versions of the functions */
#endif
```
I can `#ifdef` based on the Clang version but it doesn't help if we're afraid that this bug could exist even in Clang 8.

Reply:
This is an old bug, I'm much less concerned that it's fixed by now than that someone will use an older compiler and we don't know where in the clang release cycle the bug was fixed. On the other hand, we can only push this into the future for so long, at some point, we're going to have to move forward.

@michaelcahill, is there a reasonably easy way we could check the assembly and figure out under which compiler versions this fails?

Reply:
It's going to be hard to check the assembly: it's all moved around so much since the original bug fired.  That said, the symptom was dramatic: we were evicting pages even though the atomic change of `ref->state` to `WT_REF_LOCKED` had failed.

I'm inclined to make the change we want and run stress testing against a `clang` build for a day: given how much our test coverage has improved, if that doesn't show failure I think we're safe.

Reply:
@keithbostic @michaelcahill 
Ok, I plan to compile this change with the offending `clang` version and if it passes testing then we can probably conclude that the `__atomic` intrinsics don't suffer from the same issue.

It's difficult if it fails but we can cross that bridge if/when we come to it.

Reply:
9f3ca3b

I've added a check to stop us from compiling with `clang-3.5` or earlier.

I've run some stress testing with `clang-3.9` (that's the oldest one on our Jenkins). I've tried twice and it runs fine for ~8 hours without issues but eventually terminates with a stuck cache (this has been an issue even in `develop` for the past few days). It's a bit difficult to judge since `develop` is not passing test format stress testing right now but I'd expect that this compiler bug would cause us to fail faster than that.

@michaelcahill @keithbostic 
If you're satisfied, I think this is ready to go.